lib/gameState/GameStatePackVisitor.cpp

```diff
@@ -929,6 +929,9 @@ void GameStatePackVisitor::visitDischargeArtifact(DischargeArtifact & pack)
 		ePack.posPack.push_back(pack.artLoc.value().slot);
 		ePack.visit(*this);
 	}
+	// Workaround to inform hero bonus node about changes. Obviously this has to be done somehow differently.
+	if(pack.artLoc.has_value())
+		gs.getHero(pack.artLoc.value().artHolder)->nodeHasChanged();
```

Comment:
Calling `nodeHasChanged` on artifact itself should be enough, for example as part of `discharge` call. 

Invalidating a node will also invalidate everything located "downwards" from the node, in this case - hero & his army.

Reply:
In case of artifact node, `children` vector is empty. Even, if add hero node in it, It will work for a while but it will break after saving-loading.